scipy/stats/tests/test_hypotests.py

```diff
@@ -614,7 +614,18 @@ def test_gh_2118(self, x, y, alternative, expected):
         assert_allclose(res, expected, rtol=1e-12)
 
 
-class TestSomersD:
+class TestSomersD(_TestPythranFunc):
+    def setup_method(self):
+        self.dtypes = self.ALL_INTEGER + self.ALL_FLOAT
+        self.arguments = [np.arange(10), np.arange(10)]
+        self.index = [0, 1]
```

Comment:
Is `self.index` always (`range(len(self.arguments))`)? If so, it does not have to be specified separately.

Reply:
No, sometimes the function's arguments are not of interest, like `nonetype`, `str`, and we should not test the (float/int) dypes for such arguments. Also, it came to me that different arguments may have different dtypes to test?

Reply:
If an argument is not of interest, maybe it should not be given in this list? That can always be avoided by using `functools.partial`.

The case that I'm not sure needs to be handled is:
```
def somefunc(x0, str_param, x1):
```
That pattern is so unusual that it can be ignored I think.

> Also, it came to me that different arguments may have different dtypes to test?

That does seem more likely to happen. So maybe all this can be a single dictionary? Something like:
```
self.arguments = {0: (np.arange(10), (self.ALL_INTEGER + SELF.ALL_FLOAT)),
                  1: ...}
```

Reply:
> If an argument is not of interest, maybe it should not be given in this list? 

Yes! So I was using `self.index` to store only the index of the argument of interest(then the length is not always (range(len(self.arguments)))), and it can also handle the case `def somefunc(x0, str_param, x1):`.

> That does seem more likely to happen. So maybe all this can be a single dictionary?

Make sense! A single dictionary is absolutely better than using several lists, I'll use that :)

Reply:
Ah, no , `self.arguments` are supposed to store all the arguments needed for that function(including that not of interests), and if we used dictionary, we can't know those arguments because the key(0, 1, ...) is only those arguments of interests. Moreover, we can do `self.pythranfunc(*self.arguments)` if it is a list, otherwise we need to create a new list every time? (This is also okay I think)

Reply:
Now I switched to use a single dictionary, and I made the following changes(not sure whether it is reasonable):
- To test different dtypes for different input arrays, e.g. `def func(a,b)`, `a` have 5 dtypes to test, `b` have 3 dtypes to test.
 I don't think we should test all the `5*3` combinations of `a` dtypes and `b` dtypes because `a` and `b` would not influence each other. As long as `int` `a` & `int` `b` pass tests, `float` `a` & `float` `b` works, then `int` `a` & `float` `b` will definitely work too! (Unless there limits `a` and `b` must be the same type, but that's not what Pythran could do). Therefore, in my implementation, we only need to test `5` times rather than `5*3`. We iterate on `for idx in range(5)` types and for `b` which only has 3 types to test, it will keep using the last dtype for `idx>3`. See `get_max_dtype_list_length`, `get_dtype`, `test_all_dtypes` for details.
- To test keywords. e.g. Test Somersd `def somersd(x, y=None)`.
In my understanding, testing keywords means testing whether the results of specifying the keyword arguments(`somersd(x, None)`) vs not-specifying the keyword arguments (`somersd(x)`) are the same.
Therefore, I found it difficult to make `test_keywords` similar to `test_dtypes`/`test_views`/`test_stride`. For `def somersd(x, y=None)`, `y` can also be an array that requires testing dtypes or views, and if `y` is not None, `x` and `y` must be 1d array. If `y` is None, `x` must be a 2d array. In this way, when we `test_dtypes`/`test_views`, the input args are two 1d array, when we `test_keywords`, the input arg is one 2d array. It would be confusing if we defined so many different `self.arguments`/ `self.nonoptional_arguments` in the child class. Therefore, I choose to move `test_keywords` to the child class rather than remain in the parent class, but we provide util functions in the parent class like `self.get_optional_args` for `test_keywords` to use.